runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs

```diff
@@ -304,14 +324,22 @@ impl SnapshotStorageRebuilder {
                     .get(&old_append_vec_id)
                     .unwrap();
 
-                let storage_entry = remap_and_reconstruct_single_storage(
-                    slot,
-                    old_append_vec_id,
-                    current_len,
-                    path.as_path(),
-                    &self.next_append_vec_id,
-                    &self.num_collisions,
-                )?;
+                let storage_entry = match &self.snapshot_from {
```

Comment:
Ok, I can move "match &self.snapshot_from" out of the entry loop.  I tried the following, to define two different closures, and then just make the closure call inside the entry loop.  But somehow I got some compilation error.

```
        let process_entry = match &self.snapshot_from {
            SnapshotFrom::Archive => 
                |path: &Path,
                 current_len: usize,
                 old_appendvec_id: usize|
                 -> Result<Arc<AccountStorageEntry>, std::io::Error> {
                    remap_and_reconstruct_single_storage(
                        slot,
                        old_appendvec_id,
                        current_len,
                        path,
                        &self.next_append_vec_id,
                        &self.num_collisions,
                    )
                },

            SnapshotFrom::Dir => 
                |path: &Path,
                 current_len: usize,
                 old_appendvec_id: usize|
                -> Result<Arc<AccountStorageEntry>, std::io::Error> {
                    reconstruct_single_storage(&slot, path, current_len, old_appendvec_id as u32)
                },
        };
   ```
   
   The compilation error is
```
error[E0308]: `match` arms have incompatible types
   --> runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:322:17
    |
305 |            let process_entry = match &self.snapshot_from {
    |                                ------------------------- `match` arms have incompatible types
306 |                SnapshotFrom::Archive => 
307 |                    |path: &Path,
    |  __________________-
    | | _________________|
    | ||
308 | ||                  current_len: usize,
309 | ||                  old_appendvec_id: usize|
310 | ||                  -> Result<Arc<AccountStorageEntry>, std::io::Error> {
    | ||____________________________________________________________________- the expected closure
...   |
318 | |                      )
319 | |                  },
    | |__________________- this is found to be of type `[closure@runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:307:17: 310:69]`
...
322 |  /                 |path: &Path,
323 |  |                  current_len: usize,
324 |  |                  old_appendvec_id: usize|
325 |  |                 -> Result<Arc<AccountStorageEntry>, std::io::Error> {
326 |  |                     reconstruct_single_storage(&slot, path, current_len, old_appendvec_id as u32)
327 |  |                 },
    |  |_________________^ expected closure, found a different closure
    |
    = note: expected closure `[closure@runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:307:17: 310:69]`
               found closure `[closure@runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:322:17: 325:68]`
    = note: no two closures, even if identical, have the same type
    = help: consider boxing your closure and/or using it as a trait object
```


I see that two closures have the exactly the same input and return types.  But it seems the compiler does not accept it.  Should I box it or somehow use a trait?

Reply:
Tried, somehow boxing doesn't help.
```
    = note: expected struct `Box<[closure@runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:307:17: 310:69]>`
               found struct `Box<[closure@runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:323:17: 326:69]>`
    = note: no two closures, even if identical, have the same type
    = help: consider boxing your closure and/or using it as a trait object

Reply:
Yeah don't worry for now, like I said it's out of scope for this PR. We can come back to this once we have it working, it's just a potential for speed up, not even sure it will help in a meaningful way.